File: scanner.cpp
#include <sstream>
#include <vector>
#include <unordered_map>
#include <string>
#include <iostream>
#include <fstream>
using namespace std;


enum TokenType {
    T_AND,              // &&
    T_ASSIGN,           // =
    T_BOOLTYPE,         // bool
    T_BREAK,            // break
    T_CHARCONSTANT,     // char_lit (character literal)
    T_COMMA,            // ,
    T_COMMENT,          // comment
    T_CONTINUE,         // continue
    T_DIV,              // /
    T_DOT,              // .
    T_ELSE,             // else
    T_EQ,               // ==
    T_EXTERN,           // extern
    T_FALSE,            // false
    T_FOR,              // for
    T_FUNC,             // func
    T_GEQ,              // >=
    T_GT,               // >
    T_ID,               // identifier
    T_IF,               // if
    T_INTCONSTANT,      // int_lit (integer literal)
    T_INTTYPE,          // int
    T_LCB,              // {
    T_LEFTSHIFT,        // <<
    T_LEQ,              // <=
    T_LPAREN,           // (
    T_LSB,              // [
    T_LT,               // <
    T_MINUS,            // -
    T_MOD,              // %
    T_MULT,             // *
    T_NEQ,              // !=
    T_NOT,              // !
    T_NULL,             // null
    T_OR,               // ||
    T_PACKAGE,          // package
    T_PLUS,             // +
    T_RCB,              // }
    T_RETURN,           // return
    T_RIGHTSHIFT,       // >>
    T_RPAREN,           // )
    T_RSB,              // ]
    T_SEMICOLON,        // ;
    T_STRINGCONSTANT,   // string_lit (string literal)
    T_STRINGTYPE,       // string
    T_TRUE,             // true
    T_VAR,              // var
    T_VOID,             // void
    T_WHILE,            // while
    T_WHITESPACE,       // whitespace
    Unknown
};

struct Token{
    std::string value;
    TokenType type;
    int line_number;
    int col_start;
    int col_end;
    Token(TokenType t, const std::string& v, int l, int c1, int c2): type(t), value(v), line_number(l), col_start(c1), col_end(c2)
    {

    }
};



class LexicalAnalyzer{
    private:

        std::unordered_map<std::string, TokenType> keywords;
        std::unordered_map<std::string, TokenType> operators;
        std::unordered_map<std::string, TokenType> punctuation;
        std::unordered_map<std::string, TokenType> literals;
        vector<Token> tokens;
        string *input;
        int position;
        int line_num;
        int last_line_pos;

        void initKeywords() {
            // Keywords
            keywords["bool"] = T_BOOLTYPE;
            keywords["break"] = T_BREAK;
            keywords["continue"] = T_CONTINUE;
            keywords["else"] = T_ELSE;
            keywords["extern"] = T_EXTERN;
            keywords["false"] = T_FALSE;
            keywords["for"] = T_FOR;
            keywords["func"] = T_FUNC;
            keywords["if"] = T_IF;
            keywords["int"] = T_INTTYPE;
            keywords["null"] = T_NULL;
            keywords["package"] = T_PACKAGE;
            keywords["return"] = T_RETURN;
            keywords["string"] = T_STRINGTYPE;
            keywords["true"] = T_TRUE;
            keywords["var"] = T_VAR;
            keywords["void"] = T_VOID;
            keywords["while"] = T_WHILE;
        }

        void initOperators() {
            // Operators
            operators["&&"] = T_AND;
            operators["="] = T_ASSIGN;
            operators["=="] = T_EQ;
            operators[">="] = T_GEQ;
            operators[">"] = T_GT;
            operators["<="] = T_LEQ;
            operators["<"] = T_LT;
            operators["!="] = T_NEQ;
            operators["!"] = T_NOT;
            operators["||"] = T_OR;
            operators["+"] = T_PLUS;
            operators["-"] = T_MINUS;
            operators["*"] = T_MULT;
            operators["/"] = T_DIV;
            operators["%"] = T_MOD;
            operators["<<"] = T_LEFTSHIFT;
            operators[">>"] = T_RIGHTSHIFT;
        }

        void initPunctuation() {
            // Punctuation
            punctuation[","] = T_COMMA;
            punctuation["."] = T_DOT;
            punctuation[";"] = T_SEMICOLON;
            punctuation["{"] = T_LCB;
            punctuation["}"] = T_RCB;
            punctuation["("] = T_LPAREN;
            punctuation[")"] = T_RPAREN;
            punctuation["["] = T_LSB;
            punctuation["]"] = T_RSB;
        }

        bool isWhiteSpace(char c){ return (c == ' ' || c == '\n' || c == '\r' || c == '\t' || c == '\v' || c == 'f'); }

        bool isLetter(char c){ return (('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z')); }

        bool isDigit(char c){ return ('0' <= c && c <= '9'); }

        void removeComments(string* input) {
            int position = 0;
            int end = (*input).length();

            while (position < end) {
                if ((*input)[position] == '/' && position < end - 1 && (*input)[position] == '/') {
                    // Skip single-line comment
                    (*input)[position] = ' ';
                    (*input)[position + 1] = ' ';
                    position += 2;
                    while (position < end && (*input)[position] != '\n') {
                        (*input)[position] = ' ';
                        position++;
                    }
                } else {
                    position++;
                }
            }
        }

        int getColumnEnd(int col, string token){
            if(token.length() > 1){
                return col + token.length() - 1;
            }
            return col;
        }

        string getNextWord(int* position, string* text, int* line, int* col){
            int start;

            if(isWhiteSpace((*text)[*position])){
                while(isWhiteSpace((*text)[*position])){
                    if ((*text)[*position] == '\n') {
                        line_num++; 
                        last_line_pos = *position;
                    }
                    *position += 1;
                }
                start = *position;
            }else{
                start = *position;
            }
            
            while(!isWhiteSpace((*text)[*position])){
                *position += 1;
            }
            // Set line number and starting column number for the word
            *line = line_num;
            *col = start - last_line_pos;
            return (*text).substr(start, (*position-start));
        }

     std::string getType(TokenType type) {
            switch (type) {
                case T_AND: return "T_AND";
                case T_ASSIGN: return "T_ASSIGN";
                case T_BOOLTYPE: return "T_BOOLTYPE";
                case T_BREAK: return "T_BREAK";
                case T_CHARCONSTANT: return "T_CHARCONSTANT";
                case T_COMMA: return "T_COMMA";
                case T_COMMENT: return "T_COMMENT";
                case T_CONTINUE: return "T_CONTINUE";
                case T_DIV: return "T_DIV";
                case T_DOT: return "T_DOT";
                case T_ELSE: return "T_ELSE";
                case T_EQ: return "T_EQ";
                case T_EXTERN: return "T_EXTERN";
                case T_FALSE: return "T_FALSE";
                case T_FOR: return "T_FOR";
                case T_FUNC: return "T_FUNC";
                case T_GEQ: return "T_GEQ";
                case T_GT: return "T_GT";
                case T_ID: return "T_ID";
                case T_IF: return "T_IF";
                case T_INTCONSTANT: return "T_INTCONSTANT";
                case T_INTTYPE: return "T_INTTYPE";
                case T_LCB: return "T_LCB";
                case T_LEFTSHIFT: return "T_LEFTSHIFT";
                case T_LEQ: return "T_LEQ";
                case T_LPAREN: return "T_LPAREN";
                case T_LSB: return "T_LSB";
                case T_LT: return "T_LT";
                case T_MINUS: return "T_MINUS";
                case T_MOD: return "T_MOD";
                case T_MULT: return "T_MULT";
                case T_NEQ: return "T_NEQ";
                case T_NOT: return "T_NOT";
                case T_NULL: return "T_NULL";
                case T_OR: return "T_OR";
                case T_PACKAGE: return "T_PACKAGE";
                case T_PLUS: return "T_PLUS";
                case T_RCB: return "T_RCB";
                case T_RETURN: return "T_RETURN";
                case T_RIGHTSHIFT: return "T_RIGHTSHIFT";
                case T_RPAREN: return "T_RPAREN";
                case T_RSB: return "T_RSB";
                case T_SEMICOLON: return "T_SEMICOLON";
                case T_STRINGCONSTANT: return "T_STRINGCONSTANT";
                case T_STRINGTYPE: return "T_STRINGTYPE";
                case T_TRUE: return "T_TRUE";
                case T_VAR: return "T_VAR";
                case T_VOID: return "T_VOID";
                case T_WHILE: return "T_WHILE";
                case T_WHITESPACE: return "T_WHITESPACE";
                case Unknown: return "Unknown";
                default: return "Invalid TokenType";
            }
        }

    public:

        LexicalAnalyzer(string* source){
            input = source;
            position = 0;
            line_num = 0;
            last_line_pos = 0;
            initKeywords();
            initOperators();
            initPunctuation();
        }

        void tokenize(){
            // remove all comments
            removeComments(input);
        
            int end = (*input).length();
            TokenType type = TokenType::Unknown;
        
            while (position < end){
                // grab the next word from the file
                int line, col;
                string word = getNextWord(&position, input, &line, &col);

                // compare current word with keywords and operators
                if(keywords.find(word) != keywords.end()){
                    type = keywords[word];
                    tokens.push_back(Token(type, word, line, col, getColumnEnd(col, word)));
                }else if(operators.find(word) != operators.end()){
                    type = operators[word];
                    tokens.push_back(Token(type, word, line, col, getColumnEnd(col, word)));
                // scan each character in the word
                }else{
                    int start = 0;
                    string literal;
                    for(int i = 0; i < word.length(); i++){
                        // grab current character and cast to a string
                        string curChar;
                        curChar.push_back(word[i]);
                        if(punctuation.find(curChar) != punctuation.end()){
                            if(!literal.empty()){
                                type = TokenType::T_ID;
                                tokens.push_back(Token(type, literal, line, col, getColumnEnd(col, literal)));
                                literal.clear();
                            }
                            type = punctuation[curChar];
                            tokens.push_back(Token(type, curChar, line, col+i, getColumnEnd(col+i, curChar)));
                        }else{
                            literal += curChar;
                        }
                    }
                    // check if
                    if(!literal.empty()){
                        type = TokenType::T_ID;
                        tokens.push_back(Token(type, literal, line, col, getColumnEnd(col, literal)));
                    }
                }
            }
        }

        void printTokens(){
            for(Token t : tokens){
                cout << t.value << "     line " << t.line_number << " Cols " << t.col_start << " - " << t.col_end << " is " << getType(t.type) << endl;
            }
        }
};

void readFile(string file_name, string* source_code){
    // read file passed from std	
	ifstream file(file_name);
    if (!file)
    {
        cerr << "Error opening \"" + file_name + "\"\n";
    }

    string input;
    while(getline(file, input)){
        *source_code += "\n" + input;
    }
    
    file.close();
}

// main method to run program
int main(int argc, char *argv[]){

    string file_name = argv[1];
    string src_code;
    readFile(file_name, &src_code);
   
    LexicalAnalyzer lexer = LexicalAnalyzer(&src_code);
    lexer.tokenize();
    lexer.printTokens();

    return 0;
}

File: utils/scanner.py
import re

class TokenType:
    T_AND = '&&'
    T_ASSIGN = '='
    T_BOOLTYPE = 'bool'
    T_BREAK = 'break'
    T_CHARCONSTANT = 'char_lit'
    T_COMMA = ','
    T_COMMENT = 'comment'
    T_CONTINUE = 'continue'
    T_DIV = '/'
    T_DOT = '.'
    T_ELSE = 'T_Else'
    T_EQ = '=='
    T_EXTERN = 'extern'
    T_FALSE = 'false'
    T_FOR = 'for'
    T_FUNC = 'func'
    T_GEQ = '>='
    T_GT = '>'
    T_ID = 'identifier'
    T_IF = 'T_If'
    T_INTCONSTANT = 'T_IntConstant'
    T_INTTYPE = 'int'
    T_LCB = '{'
    T_LEFTSHIFT = '<<'
    T_LEQ = '<='
    T_LPAREN = '('
    T_LSB = '['
    T_LT = '<'
    T_MINUS = '-'
    T_MOD = '%'
    T_MULT = '*'
    T_NEQ = '!='
    T_NOT = '!'
    T_NULL = 'null'
    T_OR = '||'
    T_PACKAGE = 'package'
    T_PLUS = '+'
    T_RCB = '}'
    T_RETURN = 'T_Return'
    T_RIGHTSHIFT = '>>'
    T_RPAREN = ')'
    T_RSB = ']'
    T_SEMICOLON = ';'
    T_STRINGCONSTANT = 'T_STRINGCONSTANT'
    T_STRINGTYPE = 'string'
    T_TRUE = 'true'
    T_VAR = 'var'
    T_VOID = 'T_Void'
    T_WHILE = 'while'
    T_WHITESPACE = 'whitespace'
    Unknown = 'Unknown'

class Token:
    def __init__(self, type, value, line, col_start, col_end):
        self.type = type
        self.value = value
        self.line = line
        self.col_start = col_start
        self.col_end = col_end

class LexicalAnalyzer:
    def __init__(self, source):
        self.input = source
        self.position = 0
        self.line_num = 1
        self.last_line_pos = 0
        self.tokens = []
        self.keywords = {}
        self.operators = {}
        self.punctuation = {}
        self.init_keywords()
        self.init_operators()
        self.init_punctuation()

    def init_keywords(self):
        self.keywords = {
            "bool": TokenType.T_BOOLTYPE,
            "break": TokenType.T_BREAK,
            "continue": TokenType.T_CONTINUE,
            "else": TokenType.T_ELSE,
            "extern": TokenType.T_EXTERN,
            "false": TokenType.T_FALSE,
            "for": TokenType.T_FOR,
            "func": TokenType.T_FUNC,
            "if": TokenType.T_IF,
            "int": TokenType.T_INTTYPE,
            "null": TokenType.T_NULL,
            "package": TokenType.T_PACKAGE,
            "return": TokenType.T_RETURN,
            "string": TokenType.T_STRINGTYPE,
            "true": TokenType.T_TRUE,
            "var": TokenType.T_VAR,
            "void": TokenType.T_VOID,
            "while": TokenType.T_WHILE
        }

    def init_operators(self):
        self.operators = {
            "&&": TokenType.T_AND,
            "=": TokenType.T_ASSIGN,
            "==": TokenType.T_EQ,
            ">=": TokenType.T_GEQ,
            ">": TokenType.T_GT,
            "<=": TokenType.T_LEQ,
            "<": TokenType.T_LT,
            "!=": TokenType.T_NEQ,
            "!": TokenType.T_NOT,
            "||": TokenType.T_OR,
            "+": TokenType.T_PLUS,
            "-": TokenType.T_MINUS,
            "*": TokenType.T_MULT,
            "/": TokenType.T_DIV,
            "%": TokenType.T_MOD,
            "<<": TokenType.T_LEFTSHIFT,
            ">>": TokenType.T_RIGHTSHIFT
        }

    def init_punctuation(self):
        self.punctuation = {
            ",": TokenType.T_COMMA,
            ".": TokenType.T_DOT,
            ";": TokenType.T_SEMICOLON,
            "{": TokenType.T_LCB,
            "}": TokenType.T_RCB,
            "(": TokenType.T_LPAREN,
            ")": TokenType.T_RPAREN,
            "[": TokenType.T_LSB,
            "]": TokenType.T_RSB
        }

    def remove_comments(self):
        # Remove single-line comments
        self.input = re.sub(r'//.*', '', self.input)
        # Remove multi-line comments
        self.input = re.sub(r'/\*.*?\*/', '', self.input, flags=re.DOTALL)

    def tokenize(self):
        self.remove_comments()
        while self.position < len(self.input):
            current_char = self.input[self.position]
            if self.is_white_space(current_char):
                if current_char == '\n':
                    self.line_num += 1
                    self.last_line_pos = 0
                self.position += 1
                self.last_line_pos += 1
            elif self.is_letter(current_char):
                start_col = self.last_line_pos
                word = self.get_next_word()
                token_type = self.keywords.get(word, TokenType.T_ID)
                self.tokens.append(Token(token_type, word, self.line_num, start_col, start_col + len(word) - 1))
                self.last_line_pos += len(word)
            elif self.is_digit(current_char):
                start_col = self.last_line_pos
                number = self.get_next_word()
                self.tokens.append(Token(TokenType.T_INTCONSTANT, number, self.line_num, start_col, start_col + len(number) - 1))
                self.last_line_pos += len(number)
            elif self.is_punctuation(current_char):
                start_col = self.last_line_pos
                punct = current_char
                self.tokens.append(Token(self.punctuation[punct], punct, self.line_num, start_col, start_col))
                self.position += 1
                self.last_line_pos += 1
            elif self.is_operator(current_char):
                start_col = self.last_line_pos
                op = current_char
                self.tokens.append(Token(self.operators[op], op, self.line_num, start_col, start_col))
                self.position += 1
                self.last_line_pos += 1
            else:
                self.position += 1
                self.last_line_pos += 1

    def print_tokens(self):
        for token in self.tokens:
            print(f"{token.value}\tline {token.line} Cols {token.col_start}-{token.col_end} is {self.get_type(token.type)}")

    def is_white_space(self, c):
        return c in [' ', '\n', '\r', '\t', '\f']

    def is_letter(self, c):
        return 'a' <= c <= 'z' or 'A' <= c <= 'Z'

    def is_digit(self, c):
        return '0' <= c <= '9'

    def is_punctuation(self, c):
        return c in self.punctuation

    def is_operator(self, c):
        return c in self.operators

    def get_next_word(self):
        start = self.position
        while self.position < len(self.input) and (self.is_letter(self.input[self.position]) or self.is_digit(self.input[self.position])):
            self.position += 1
        return self.input[start:self.position]

    def get_type(self, type):
        return {
            TokenType.T_AND: "T_AND",
            TokenType.T_ASSIGN: "T_ASSIGN",
            TokenType.T_BOOLTYPE: "T_BOOLTYPE",
            TokenType.T_BREAK: "T_BREAK",
            TokenType.T_CHARCONSTANT: "T_CHARCONSTANT",
            TokenType.T_COMMA: "T_COMMA",
            TokenType.T_COMMENT: "T_COMMENT",
            TokenType.T_CONTINUE: "T_CONTINUE",
            TokenType.T_DIV: "T_DIV",
            TokenType.T_DOT: "T_DOT",
            TokenType.T_ELSE: "T_ELSE",
            TokenType.T_EQ: "T_EQ",
            TokenType.T_EXTERN: "T_EXTERN",
            TokenType.T_FALSE: "T_FALSE",
            TokenType.T_FOR: "T_FOR",
            TokenType.T_FUNC: "T_FUNC",
            TokenType.T_GEQ: "T_GEQ",
            TokenType.T_GT: "T_GT",
            TokenType.T_ID: "T_ID",
            TokenType.T_IF: "T_IF",
            TokenType.T_INTCONSTANT: "T_INTCONSTANT",
            TokenType.T_INTTYPE: "T_INTTYPE",
            TokenType.T_LCB: "T_LCB",
            TokenType.T_LEFTSHIFT: "T_LEFTSHIFT",
            TokenType.T_LEQ: "T_LEQ",
            TokenType.T_LPAREN: "T_LPAREN",
            TokenType.T_LSB: "T_LSB",
            TokenType.T_LT: "T_LT",
            TokenType.T_MINUS: "T_MINUS",
            TokenType.T_MOD: "T_MOD",
            TokenType.T_MULT: "T_MULT",
            TokenType.T_NEQ: "T_NEQ",
            TokenType.T_NOT: "T_NOT",
            TokenType.T_NULL: "T_NULL",
            TokenType.T_OR: "T_OR",
            TokenType.T_PACKAGE: "T_PACKAGE",
            TokenType.T_PLUS: "T_PLUS",
            TokenType.T_RCB: "T_RCB",
            TokenType.T_RETURN: "T_RETURN",
            TokenType.T_RIGHTSHIFT: "T_RIGHTSHIFT",
            TokenType.T_RPAREN: "T_RPAREN",
            TokenType.T_RSB: "T_RSB",
            TokenType.T_SEMICOLON: "T_SEMICOLON",
            TokenType.T_STRINGCONSTANT: "T_STRINGCONSTANT",
            TokenType.T_STRINGTYPE: "T_STRINGTYPE",
            TokenType.T_TRUE: "T_TRUE",
            TokenType.T_VAR: "T_VAR",
            TokenType.T_VOID: "T_VOID",
            TokenType.T_WHILE: "T_WHILE",
            TokenType.T_WHITESPACE: "T_WHITESPACE",
            TokenType.Unknown: "Unknown"
        }.get(type, "Invalid TokenType")